operator/internal/decommissioning/statefulset_decomissioner.go

```diff
@@ -0,0 +1,753 @@
+// ready to decommission due to a stale read on either the StatefulSet or a partial list of Pods. Consider:
+//
+// - A StatefulSet scales up and the pod comes up and registers with the cluster but has not yet become fully healthy
+// - Reconciliation is triggered due to the pod coming online but the StatefulSet has been slow to propagate through
+// cache
+// - The requested replicas for the StatefulSet are out-of-date and it seems like we have one extra broker than we need
+// and that broker is currently unhealthy in our cluster health check
+// - We mark it as needing to be decommissioned even though it's new
+//
+// These types of stale cache scenarios are effectively eliminated by introducing a "delayed" cache that essentially
+// counts the number of times that we've marked a broker as ready to decommission across a certain window. If it's been
+// marked as needing to be decommissioned some m times with delays of n in between each mark, then it can be decommissioned
+// because we've given the client-side cache enough time to fill. We also add this guard for PVCs. In every loop through
+// reconciliation we expunge the broker and/or PVC entries that no longer meet the decommission criteria, so if a broker
+// or PVC should no longer be decommissioned, we reset our count.
+func (s *StatefulSetDecomissioner) Decommission(ctx context.Context, set *appsv1.StatefulSet) (bool, error) {
+	// note that this is best-effort, the decommissioning code needs to be idempotent and deterministic
+
+	log := ctrl.LoggerFrom(ctx, "namespace", set.Namespace, "name", set.Name).WithName("StatefulSetDecommissioner.Decomission")
+
+	// if helm is not managing it, move on.
+	if managedBy, ok := set.Labels[k8sManagedByLabelKey]; managedBy != "Helm" || !ok {
+		log.V(traceLevel).Info("not managed by helm")
+		return false, nil
+	}
+
+	keep, err := s.filter(ctx, set)
+	if err != nil {
+		log.Error(err, "error filtering StatefulSet")
+		return false, err
+	}
+
+	if !keep {
+		log.V(traceLevel).Info("skipping decommission, StatefulSet filtered out")
+		return false, nil
+	}
+
+	unboundVolumeClaims, err := s.findUnboundVolumeClaims(ctx, set)
+	if err != nil {
+		log.Error(err, "error finding unbound PersistentVolumeClaims")
+		return false, err
+	}
+
+	log.V(traceLevel).Info("fetched unbound volume claims", "claims", functional.MapFn(func(claim *corev1.PersistentVolumeClaim) string {
+		return claim.Name
+	}, unboundVolumeClaims))
+
+	setCacheKey := client.ObjectKeyFromObject(set)
+
+	// remove any volumes from the cache that are no longer considered unbound
+	s.delayedVolumeCache.Filter(setCacheKey, functional.MapFn(func(claim *corev1.PersistentVolumeClaim) types.NamespacedName {
+		return client.ObjectKeyFromObject(claim)
+	}, unboundVolumeClaims)...)
+
+	// first mark all of the claims as needing potential expiration
+	for _, claim := range unboundVolumeClaims {
+		s.delayedVolumeCache.Mark(setCacheKey, client.ObjectKeyFromObject(claim))
+	}
```

Comment:
Why mark is called after filter?

Reply:
Maybe my understanding is broken but `Filter` deletes from the delayed volume cache all entries that would be processed later (`Mark` and `Process`)
```go
// Filter keeps only the entries in the cache which are equal to the given entries.
func (d *CategorizedDelayedCache[Category, Entry]) Filter(category Category, entries ...Entry) {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	if stored, ok := d.data[category]; ok {
		for entry := range stored {
			if !slices.Contains(entries, entry) {
				delete(stored, entry)
			}
		}
	}
}
```

Reply:
Sorry, I didn't see negation before `slices.Contains`.